include/lbann/layers/layer.hpp

```diff
@@ -209,182 +216,267 @@ class Layer {
+  /** Expected number of child layers.
    *  A negative value indicates no limit.
    */
-  int m_max_num_child_layers;
+  int m_expected_num_child_layers = 1;
 
-  model *m_model;
+  /** Reference to model managing this layer. */
+  model *m_model = nullptr;
 
-  /** Setup views of the matrices for the layer's forward propagation. */
-  virtual void fp_set_std_matrix_view();
-  /** Setup views of the matrices for the layer's backward propagation. */
-  virtual void bp_set_std_matrix_view();
+  /** Setup data for forward propagation.
+   *  Base method gets previous activations from parent layers and
+   *  resizes activations for the current mini-batch size.
+   */
+  virtual void fp_setup_data(int mini_batch_size);
+  /** Setup data for forward propagation.
+   *  Base method gets previous error signals from child layers. The
+   *  error signals are resized for the current mini-batch size in the
+   *  clear_error_signals function.
+   */
+  virtual void bp_setup_data(int mini_batch_size);
 #ifdef LBANN_HAS_CUDNN
   /** Pin host memory if needed for GPU memory transfers. */
   virtual void pin_data();
 #endif // LBANN_HAS_CUDNN
 
   /** Setup pointers to parent and child layers.
-   *  Called by the setup function. This base method just checks that
-   *  the number of parents and children are valid. Pointers to the
-   *  parent/child layers are assumed to be initialized already.
+   *  Called by the setup function. The base method checks that the
+   *  number of parents and children are valid. Pointers to the
+   *  parent/child layers are assumed to be already initialized.
    */
   virtual void setup_pointers();
-  /** Setup neuron tensor dimensions
-   *  Called by the setup function. This base method initializes the
-   *  input neuron tensor dimensions and sets the output neuron tensor
-   *  dimensions equal to the input.
+  /** Setup tensor dimensions
+   *  Called by the setup function. The base method sets the
+   *  dimensions of the activation tensors equal to the dimensions of
+   *  the first previous activation tensor.
    */
   virtual void setup_dims();
+  /** Instantiate distributed matrices.
+   *  If the layer has already been setup, this function should
+   *  destroy all matrices and reinstantiate them. However, it is not
+   *  guaranteed that derived classes will obey this behavior.
+   */
+  virtual void setup_matrices(const El::Grid& grid);
   /** Setup layer data.
-   *  Called by the setup function. This base method initializes the
-   *  activations and error signal matrices.
+   *  Called by the setup function. The base method sets the previous
+   *  activation, activation, previous error signal, and error signal
+   *  matrices to zero matrices with the proper dimensions. Matrix
+   *  buffers are pinned if needed for GPU transfers.
    */
   virtual void setup_data();
+  /** Setup matrix views.
+   *  Called by the setup function. The base method does nothing.
+   */
+  virtual void setup_views() {}
```

Comment:
Does this function still need to exist now that we have removed matrix views?

Reply:
That's a good point. The setup_data function has subsumed its functionality.